core/state_processor.go

```diff
@@ -151,10 +157,32 @@ func ApplyTransaction(config *params.ChainConfig, bc ChainContext, author *commo
 
 	var cxReceipt *types.CXReceipt
 	if txType == types.SubtractionOnly {
-		cxReceipt = &types.CXReceipt{tx.Hash(), msg.Nonce(), msg.From(), msg.To(), tx.ShardID(), tx.ToShardID(), msg.Value()}
+		cxReceipt = &types.CXReceipt{tx.Hash(), msg.From(), msg.To(), tx.ShardID(), tx.ToShardID(), msg.Value()}
 	} else {
 		cxReceipt = nil
 	}
 
 	return receipt, cxReceipt, gas, err
 }
+
+// ApplyIncomingReceipt will add amount into ToAddress in the receipt
+func ApplyIncomingReceipt(config *params.ChainConfig, db *state.DB, header *types.Header, cxp *types.CXReceiptsProof) {
+	if cxp == nil {
+		return
+	}
+
+	// TODO: how to charge gas here?
+	for _, cx := range cxp.Receipts {
+		if cx == nil || cx.To == nil { // should not happend
+			utils.Logger().Warn().Msg("ApplyIncomingReceipts: Invalid incoming receipt!!")
+			continue
+		}
+		utils.Logger().Info().Msgf("ApplyIncomingReceipts: ADDING BALANCE %d", cx.Amount)
+
+		if !db.Exist(*cx.To) {
+			db.CreateAccount(*cx.To)
+		}
+		db.AddBalance(*cx.To, cx.Amount)
+		db.IntermediateRoot(config.IsEIP158(header.Number)).Bytes()
```

Comment:
Should we be consistent with ApplyTransaction:
        if config.IsByzantium(header.Number) {
		statedb.Finalise(true)
	} else {
		root = statedb.IntermediateRoot(config.IsEIP158(header.Number)).Bytes()
	}

Reply:
Maybe, I am not quite sure about the byzantium block. But I think we should remove the isByzantium block (sadly we already passed that block number in ethereum). Let me know what do you think

Reply:
we haven't "ByzantiumBlock:      big.NewInt(4370000)," it's 4.3 million blocks. Our single shard blocks haven't go beyond 1m

Reply:
yeah, wrong memory, we passed the EIP155 block. In this case, we could delete the Byzantium block checking here

Reply:
I checked the code, here it doesn't matter, both will use finalise(true). However, Byzantium fork contains several security improvement. I would suggest to make the byzantium block num smaller to benefit from that.